CIP-0112? | Observe script type (#749)
cardano-foundation/CIPs: cip-observe-script-type/README.md

---

> CIP: ?

Reviewer: ```suggestion
CIP: 112
```

Title: Observe Script Type
Status: Proposed
Category: Plutus
Authors:
    - Philip DiSarro <info@anastasialabs.com>
Implementors: []
Discussions: 
    - https://github.com/cardano-foundation/CIPs/pull/418/

> Created: 2024-1-8

Reviewer 2: ```suggestion
Created: 2024-01-08
```
fix date format :)

This proposal makes this design pattern indepedent from implementation details of stake validators and withdrawals, and improves efficiency and readability for validators that implement it. 

## Specification
<!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. This must include how the CIP should be versioned. If a proposal defines structure of on-chain data it must include a CDDL schema in it's specification.-->
The type signature of this script type will be consistent with the type signature of minting and staking validators, namely:
```haskell
Redeemer -> ScriptContext -> () 
```

The type signature of the newly introduced `Purpose` will be:
```haskell
Observe Integer -- ^ where integer is the index into the observations list.
```

### Script context

> Scripts are passed information about transactions via the script context.
> We propose to augment the script context to include information about the observation scripts that are executed in the transaction.
> 
> Changing the script context will require a new Plutus language version in the ledger to support the new interface.
> The change is: a new field is added to the script context which represents the list of observation scripts that must validate the transaction.
> The observation scripts in the list are represented by their hash. 
> 
> The interface for old versions of the language will not be changed.
> Scripts with old versions cannot be spent in transactions that include observation scripts, attempting to do so will be a phase 1 transaction validation failure.
> 

Reviewer 3: ```suggestion
```

## Rationale: how does this CIP achieve its goals?
<!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It should describe alternate designs considered and related work. The rationale should provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.

It must also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section should explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.
-->

## Path to Active

### Acceptance Criteria
<!-- Describes what are the acceptance criteria whereby a proposal becomes 'Active' -->
- [] Fully implemented in Cardano.
      
### Implementation Plan
<!-- A plan to meet those criteria. Or `N/A` if not applicable. -->
- [] Passes all requirements of both Plutus and Ledger teams as agreed to improve Plutus script efficiency and usability.
      
## Copyright
<!-- The CIP must be explicitly licensed under acceptable copyright terms. -->

> 

Reviewer 2: ```suggestion
This CIP is licensed under [CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/legalcode).

```
to match license in the header

then it becomes **O(obs_N)** where `obs_N` is the number of Observe validators that are executed in the transaction as you have to verify that the StakingCredential is present in `txInfoWdrl`.

This proposal makes this design pattern indepedent from implementation details of stake validators and withdrawals, and improves efficiency and readability for validators that implement it. 

## Specification
<!-- The technical specification should describe the proposed improvement in sufficient technical detail. In particular, it should provide enough information that an implementation can be performed solely on the basis of the design in the CIP. This is necessary to facilitate multiple, interoperable implementations. This must include how the CIP should be versioned. If a proposal defines structure of on-chain data it must include a CDDL schema in it's specification.-->
The type signature of this script type will be consistent with the type signature of minting and staking validators, namely:
```haskell
Redeemer -> ScriptContext -> () 
```

The type signature of the newly introduced `Purpose` will be:
```haskell
Observe Integer -- ^ where integer is the index into the observations list.

> ```

Reviewer 4: Is there a reason why the purpose should include the index instead of the script hash like with minting policies and staking scripts? The spending purpose is the only one that does not provide the hash directly and I only ever use it to lookup the script's hash from the input being spent. It seems other developers do the same since there is now even this [CIP](https://github.com/cardano-foundation/CIPs/pull/754) which suggests changing the spending purpose to include the script's hash. I would much rather have the `Purpose` be:
``` Haskell
Observe ScriptHash -- ^ the hash of the script currently observing.
```

Author: I would argue it should be the other way around, the spending purpose should include the index of the UTxO being spent. The standard usage for spending scripts is to find the input that is currently being validated against.

Reviewer 4: > the spending purpose should include the index of the UTxO being spent.

I think this makes sense for spending scripts but not for observation scripts. A `TxInInfo` contains a lot more information than just a `TxOutRef` or a `ScriptHash`. It probably doesn't make sense to put *all* of that information into the `Purpose` so an index place-holder makes sense. But for observation scripts, the only thing you can get from the `TxInfoObservations` is a `ScriptHash`, and a `ScriptHash` fits no problem in the `Purpose`. If the observation index serves no purpose other than to get the script hash for the current observation script, then doesn't it makes more sense to inline the script hash instead of using an index place-holder? Otherwise, the index is forcing the unnecessary extra step of looking up the script hash in the observation list. Do you personally have another use for the observation index in mind?

Author: Yes, I agree. 

Spending Script purpose should be `Spending Integer ScriptHash` and observe purpose should be `Observe ScriptHash`

Reviewer 3: This is a weak spot in the current design of things to be sure. The thing that would be closest to the ledger would be to always provide a "pointer" to the thing that caused the run, in which case we should indeed probably change the `Spending` purpose to have an index instead. The current situation is a bit inconsistent; in some cases we resolve some of the information for you, in others not.

Author: If we do decide to resolve information for ScriptPurpose, then we should make the information we resolve consistent across all the different constructors. I think `ScriptHash` is a good option in that case, since it is universally useful across the common script types (Minting, Spending, Staking). 

But I really like the idea of integer indices for everything because then we can move towards a future where we can work towards accessing the unique efficiency benefits from transaction determinism that other smart contract platforms (ie EVM) cannot have access to (especially if in the far future we have support for constant access indexed data structures directly in the script context). I feel like currently, we haven't even begun to scratch the surface of utilizing the advantages afforded to us via Cardano's transaction determinism. 

For instance, since we know at the time of transaction construction what the majority of the script context will look like, we can theoretically bring most lookups from O(n) to O(1). Especially with [TxOut Redeemers](https://github.com/cardano-foundation/CIPs/pull/735) you can imagine a world where common lookup operations like `findOwnInput`, `valueOf`, `findStateThreadOutput`, are all O(1). Instead of looking up elements on-chain, we look them up off-chain and provide the index to the on-chain computation which then only has to check that the element is indeed where we claim it is.

This change is not backwards-compatible and will need to go into a new Plutus language version.

### Alternatives 

- We could decide to accept the withdraw-zero staking script trick as an adequate solution, and just preserve the nonsensical withdraw zero case in future language versions. 
- The staking script trick could be abstracted away from the developer by smart contract languages that compile to UPLC. 
    - This can be dangerous since by distancing the developer from what is actually happening you open up the door for the developer to act on misguided assumptions. 

## Path to Active

### Acceptance Criteria

> - [] Fully implemented in Cardano.

Reviewer 2: ```suggestion
- [ ] These rules included within a official Plutus version, and released via a major hard fork.
```

      
### Implementation Plan

> - [] Passes all requirements of both Plutus and Ledger teams as agreed to improve Plutus script efficiency and usability.

Reviewer: p.s. to https://github.com/cardano-foundation/CIPs/pull/749#discussion_r1585200445 ... changing the other tickbox to proper GitHub format. 🤓
```suggestion
- [ ] Passes all requirements of both Plutus and Ledger teams as agreed to improve Plutus script efficiency and usability.
```

---
CIP: 112
Title: Observe Script Type
Status: Proposed
Category: Plutus
Authors:
    - Philip DiSarro <info@anastasialabs.com>
Implementors: []
Discussions: 
    - https://github.com/cardano-foundation/CIPs/pull/418/
Created: 2024-01-08
License: CC-BY-4.0
---

> 
> <!-- Existing categories:
> 
> - Meta     | For meta-CIPs which typically serves another category or group of categories.
> - Wallets  | For standardisation across wallets (hardware, full-node or light).
> - Tokens   | About tokens (fungible or non-fungible) and minting policies in general.
> - Metadata | For proposals around metadata (on-chain or off-chain).
> - Tools    | A broad category for ecosystem tools not falling into any other category.
> - Plutus   | Changes or additions to Plutus
> - Ledger   | For proposals regarding the Cardano ledger (including Reward Sharing Schemes)
> - Catalyst | For proposals affecting Project Catalyst / the Jörmungandr project
> 
> -->
> 

Reviewer: ```suggestion
```
... etc [in all sections] ...

  , txInfoSignatories           :: [V2.PubKeyHash]
  , txInfoRedeemers             :: Map ScriptPurpose V2.Redeemer
  , txInfoData                  :: Map V2.DatumHash V2.Datum
  , txInfoId                    :: V2.TxId
  , txInfoVotingProcedures      :: Map Voter (Map GovernanceActionId VotingProcedure)
  , txInfoProposalProcedures    :: [ProposalProcedure]
  , txInfoCurrentTreasuryAmount :: Haskell.Maybe V2.Value
  , txInfoTreasuryDonation      :: Haskell.Maybe V2.Value
  , txInfoObservations          :: [ScriptHash] -- ^ newly introduced list of observation scripts that executed in this tx. 
  }
```

### CDDL

The CDDL for transaction body will change as follows to reflect the new field.
```
transaction_body =
  { 0 : set<transaction_input>             ; inputs
  , 1 : [* transaction_output]
  , 2 : coin                               ; fee
  , ? 3 : uint                             ; time to live
  , ? 4 : certificates
  , ? 5 : withdrawals
  , ? 7 : auxiliary_data_hash
  , ? 8 : uint                             ; validity interval start
  , ? 9 : mint
  , ? 11 : script_data_hash
  , ? 13 : nonempty_set<transaction_input> ; collateral inputs

>   , ? 14 : required_signers
>   , ? 15 : network_id
>   , ? 16 : transaction_output              ; collateral return
>   , ? 17 : coin                            ; total collateral
>   , ? 18 : nonempty_set<transaction_input> ; reference inputs
>   , ? 19 : voting_procedures               ; Voting procedures
>   , ? 20 : proposal_procedures             ; Proposal procedures
>   , ? 21 : coin                            ; current treasury value
>   , ? 22 : positive_coin                   ; donation
>   , ? 23 : required_observers              ; New; observation scripts that must execute in phase 2 validation
>   }
> 
> required_observers = set<scripthash>

Reviewer 5: I would suggest to rather have `required_signers` field be promoted to the more powerful `required_observers`. The nice thing about it we can do this in a backwards compatible manner in order to allow tooling for graceful transition:
```suggestion
  , ? 14 : required_observers              ; Upgraded `required_signers`
  , ? 15 : network_id
  , ? 16 : transaction_output              ; collateral return
  , ? 17 : coin                            ; total collateral
  , ? 18 : nonempty_set<transaction_input> ; reference inputs
  , ? 19 : voting_procedures               ; Voting procedures
  , ? 20 : proposal_procedures             ; Proposal procedures
  , ? 21 : coin                            ; current treasury value
  , ? 22 : positive_coin                   ; donation
  }

; addr_keyhash variant is included for backwards compatibility and will be 
; deprecated in the future era, because `credential` already contains `addr_keyhash`.
required_observers = nonempty_set<credential / addr_keyhash>
```

Reviewer 5: > I don't think there's a need to change the field name 

We'd have to rename the field, since it's type in ledger would change.

That will not work, since it is not backwards compatible with the previous format
```
required_signers = [ * credential ]
```

```

> The `required_observers` (field 23) is a set of script hashes that can be used to require the associated Plutus script to be present in the witness set or as a reference script and executed in the transaction. If a script hash is present but the corresponding Plutus script is not in the witness set or present in a reference script, the transaction will fail in phase 1 validation. This way plutus scripts can check the script context to know which observation scripts were executed in the transaction.

Reviewer 4: > The `required_observers` (field 23) is a set of script hashes that can be used to require the associated Plutus script ...

Native scripts should also be in the `required_observers` list. It should be:
> ... require the associated script ...

Reviewer 4: Unless I am wrong (if I am, please correct me) and there is way for plutus scripts to verify that a native script was executed **without** having to look for side-effects like spending/minting/withdrawing, then native scripts should also be in the observe list. This way plutus scripts can verify a native script was executed *and* native scripts can verify a plutus script was executed; all without having to rely on ledger side-effects. Multisig is a really good use-case for this.

           | <RequireObserver>   <scripthash>
           | <RequireTimeBefore> <slotno>
           | <RequireTimeAfter>  <slotno>

           | <RequireAllOf>      <native_script>*
           | <RequireAnyOf>      <native_script>*
           | <RequireMOf>        <num> <native_script>*
```

Native scripts are typically represented in JSON syntax. We propose the following JSON representation for the `RequireObserver` constructor:
```JSON
{
  "type": "observer",
  "keyHash": "OBSERVATION_SCRIPT_HASH"
}
```

## Rationale: how does this CIP achieve its goals?
<!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It should describe alternate designs considered and related work. The rationale should provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.

It must also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section should explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.
-->

> Currently Plutus scripts in a transaction will only execute when the transaction performs the associated ledger action (ie. a Plutus minting policy will only execute if the transaction mints or burns tokens with matching currency symbol). The only exception is the withdraw zero trick which relies on an obscure mechanic where zero amount withdrawals are not filtered by the ledger. Now using `required_observers` we can specify a list of Plutus scripts to be executed in the transaction independent of any ledger actions. The newly introduced `txInfoObservations` field in the script context provides a straightforward way for Plutus scripts to check that "a particular script validated this transaction".

Reviewer 4: Here, too, it focuses exclusively on plutus scripts.
> Currently Plutus scripts in a transaction will only execute when the transaction performs the associated ledger action ...

The same is true for native scripts.

> This change is not backwards-compatible and will need to go into a new Plutus language version.

Reviewer 4: I'm not sure what the equivalent of a `plutus language version` is for native scripts but I am guessing they will also need a new version.

<!-- A short (\~200 word) description of the proposed solution and the technical issue being addressed. -->
We propose to introduce a new Plutus scripts type `Observe` in addition to those currently available (spending, certifying, rewarding, minting, drep). The purpose of this script type is to allow arbitrary validation logic to be decoupled from any ledger action. 
Since observe validators are decoupled from actions, you can run them in a transaction without needing to perform any associated action (ie you don't need to consume a script input, or mint a token, or withdraw from a staking script just to execute this validator). 

> Additionally, we propose to introduce a new assertion to native scripts that they can use to check that a particular script hash is in `required_observers` (which in turn enforces that the script must be executed successfully in the transaction). This addresses a number of technical issues discussed in other CIPs and CPS such as the redundant execution of spending scripts, and the inability to effectively use native scripts in conjunction with Plutus scripts. 

Reviewer 6: If we're making a change to the native script version, we should probably also get things on POSIX time rather than slot times.

Reviewer 6: Whether that's another CIP or part of this one, i'd probably lean to a separate CIP, but it'd be a shame to see one without the other.
